x/upgrade/internal/keeper/keeper.go

```diff
@@ -130,3 +140,66 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
 func (k Keeper) IsSkipHeight(height int64) bool {
 	return k.skipUpgradeHeights[height]
 }
+
+// DumpUpgradeInfoToDisk adds plan height to upgrade-info.json
+func (k Keeper) DumpUpgradeInfoToDisk(height int64, name string) error {
+	upgradeInfoFilePath, err := k.GetUpgradeInfoPath()
+	if err != nil {
+		return err
+	}
+
+	upgradeInfo := store.UpgradeInfo{
+		Name:   name,
+		Height: height,
+	}
+	info, err := json.Marshal(upgradeInfo)
+
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(upgradeInfoFilePath, info, 0644)
+}
+
+// GetUpgradeInfoPath returns the upgrade info file path
+func (k Keeper) GetUpgradeInfoPath() (string, error) {
+	upgradeInfoFileDir := path.Join(k.GetHomeDir(), "data")
+
+	if _, err := os.Stat(upgradeInfoFileDir); os.IsNotExist(err) {
+		err = os.Mkdir(upgradeInfoFileDir, os.ModePerm)
+		if err != nil {
+			return "", err
+		}
+		return filepath.Join(upgradeInfoFileDir, UpgradeInfoFileName), nil
+	}
+
+	return filepath.Join(upgradeInfoFileDir, UpgradeInfoFileName), nil
+}
+
+// GetHomeDir returns the height at which the given upgrade was executed
+func (k Keeper) GetHomeDir() string {
+	return k.homePath
+}
+
+// ReadUpgradeInfoFromDisk returns the name and height of the upgrade
+// which is written to disk by the old binary when panic'ing
+// if there's an error in reading the info,
+// it assumes that the upgrade info is not available
+func (k Keeper) ReadUpgradeInfoFromDisk() (upgradeName string, upgradeHeight int64) {
+	upgradeInfoPath, err := k.GetUpgradeInfoPath()
+	// if error in reading the path, assume there are no upgrades
+	if err != nil {
+		return "", -1
+	}
+
+	data, err := ioutil.ReadFile(upgradeInfoPath)
+
+	// if error in reading the file, assume there are no upgrades
+	if err != nil {
+		return "", -1
+	}
+
+	var upgradeInfo store.UpgradeInfo
+	json.Unmarshal(data, &upgradeInfo)
+	return upgradeInfo.Name, upgradeInfo.Height
```

Comment:
You're returning all the values from the `UpgradeInfo`  type. You might as well just return the entire `UpgradeInfo` struct which will cleanup the method signature 👍

Reply:
Updated the required changes